### DISCUSSION TITLE
How can I extract a list of IDs from winget upgrade?

### DISCUSSION THREAD

**Discussion by BluesBroBo** [microsoft/winget-cli#2589] [PowerShell]
I'm trying to make a powershell variable that contains the IDs of all software that has updates available.
However something seems to be wrong with the way winget outputs the Updates it finds. When I select-object the IDs I just get the Title Line and a bunch of empty lines.
Is there another way to do this?

**Comment by jantari** [comment]
The problem you're seeing is due to the incomprehensive fact that winget is not actually a PowerShell command (that outputs structured PowerShell data objects) but rather a simple DOS executable that just outputs plain text - similar to `xcopy`, `ping`, `pnputil`, `netsh` etc.

There is a lot of existing discussion about this here:

https://github.com/microsoft/winget-cli/issues/221

There hasn't been an update onthe matter in a while, but last we heard about it:

https://github.com/microsoft/winget-cli/discussions/2300
https://github.com/microsoft/winget-cli/discussions/2063

retroactively bringing winget into the 21th century was progressing along and may have an initial release around winget v1.4 - v.1.5 ?

**Reply by jantari** [reply]
btw @BluesBroBo with the [`Microsoft.Winget.Client` PowerShell module v1.6](https://www.powershellgallery.com/packages/Microsoft.WinGet.Client/1.6.3133.0) this can now be done like this:

```powershell
$installed = Get-WinGetPackage -Source winget
$updatable = $installed | Where-Object IsUpdateAvailable | Select-Object -ExpandProperty Id
$updatable
```

**Reply by naldodj** [reply]
```powershell
# Limpa a tela para melhor legibilidade
Clear-Host

# Verifica se o módulo Microsoft.WinGet.Client está instalado e atualizado
if (Get-Module -ListAvailable -Name Microsoft.WinGet.Client) {
    Write-Host "Microsoft.WinGet.Client Module is installed."
    $moduleVersion = (Get-Module -Name Microsoft.WinGet.Client).Version
    if ($moduleVersion -lt [Version]"1.6.3133.0") {
        Write-Host "Updating Microsoft.WinGet.Client Module..."
        Update-Module -Name Microsoft.WinGet.Client -RequiredVersion 1.6.3133.0
    } else {
        Write-Host "Microsoft.WinGet.Client Module is up to date."
    }
} else {
    Write-Host "Microsoft.WinGet.Client Module is not installed. Installing..."
    Install-Module -Name Microsoft.WinGet.Client -RequiredVersion 1.6.3133.0 -Force
}

# Executa o comando winget upgrade para verificar atualizações disponíveis
Write-Host "`nChecking for available package updates..."
winget upgrade -h --all

# Obtém a lista de pacotes instalados e identifica os que têm atualizações disponíveis
$installed = Get-WinGetPackage -Source winget
$updatable = $installed | Where-Object IsUpdateAvailable | Select-Object -ExpandProperty Id

# Se houver pacotes atualizáveis, atualiza-os
if ($updatable) {
    Write-Host "`nUpdating available packages...`n"
    foreach ($packageId in $updatable) {
        Write-Host "Updating package: $packageId"
        winget update $packageId
		Write-Host ""
    }
} else {
    Write-Host "`nNo updates available for installed packages."
}
```
